src/entity/object/AreaEffectCloud.php

```diff
@@ -0,0 +1,384 @@
+		$this->initialRadius = $nbt->getFloat(self::TAG_INITIAL_RADIUS, self::RADIUS);
+		$this->radius = $nbt->getFloat(self::TAG_RADIUS, self::RADIUS);
+		$this->radiusChangeOnPickup = $nbt->getFloat(self::TAG_RADIUS_CHANGE_ON_PICKUP, self::RADIUS_CHANGE_ON_PICKUP);
+		$this->radiusOnUse = $nbt->getFloat(self::TAG_RADIUS_ON_USE, self::RADIUS_ON_USE);
+		$this->radiusPerTick = $nbt->getFloat(self::TAG_RADIUS_PER_TICK, self::RADIUS_PER_TICK);
+
+		/** @var CompoundTag[]|ListTag|null $effectsTag */
+		$effectsTag = $nbt->getListTag(self::TAG_EFFECTS);
+		if($effectsTag !== null){
+			foreach($effectsTag as $e){
+				$effect = EffectIdMap::getInstance()->fromId($e->getByte("Id"));
+				if($effect === null){
+					continue;
+				}
+
+				$this->effectContainer->add(new EffectInstance(
+					$effect,
+					$e->getInt("Duration"),
+					Binary::unsignByte($e->getByte("Amplifier")),
+					$e->getByte("ShowParticles", 1) !== 0,
+					$e->getByte("Ambient", 0) !== 0
+				));
+			}
+		}
+	}
+
+	public function saveNBT() : CompoundTag{
+		$nbt = parent::saveNBT();
+
+		$nbt->setLong(self::TAG_SPAWN_TICK, $this->getWorld()->getTime() - $this->age);
+		$nbt->setShort(self::TAG_POTION_ID, PotionTypeIdMap::getInstance()->toId($this->potionType));
+		$nbt->setInt(self::TAG_DURATION, $this->duration);
+		$nbt->setInt(self::TAG_DURATION_ON_USE, $this->durationOnUse);
+		$nbt->setInt(self::TAG_PICKUP_COUNT, $this->pickupCount);
+		$nbt->setInt(self::TAG_REAPPLICATION_DELAY, $this->reapplicationDelay);
+		$nbt->setFloat(self::TAG_INITIAL_RADIUS, $this->initialRadius);
+		$nbt->setFloat(self::TAG_RADIUS, $this->radius);
+		$nbt->setFloat(self::TAG_RADIUS_CHANGE_ON_PICKUP, $this->radiusChangeOnPickup);
+		$nbt->setFloat(self::TAG_RADIUS_ON_USE, $this->radiusOnUse);
+		$nbt->setFloat(self::TAG_RADIUS_PER_TICK, $this->radiusPerTick);
+
+		if(count($this->effectContainer->all()) > 0){
+			$effects = [];
+			foreach($this->effectContainer->all() as $effect){
+				$effects[] = CompoundTag::create()
+					->setByte("Id", EffectIdMap::getInstance()->toId($effect->getType()))
+					->setByte("Amplifier", Binary::signByte($effect->getAmplifier()))
+					->setInt("Duration", $effect->getDuration())
+					->setByte("Ambient", $effect->isAmbient() ? 1 : 0)
+					->setByte("ShowParticles", $effect->isVisible() ? 1 : 0);
+			}
+			$nbt->setTag(self::TAG_EFFECTS, new ListTag($effects));
+		}
+
+		return $nbt;
+	}
+
+	public function isFireProof() : bool{
+		return true;
+	}
+
+	public function canBeCollidedWith() : bool{
+		return false;
+	}
+
+	public function getAge() : int{
+		return $this->age;
+	}
+
+	public function getPotionType() : PotionType{
+		return $this->potionType;
+	}
+
+	public function getEffects() : EffectContainer{
+		return $this->effectContainer;
+	}
+
+	public function getInitialRadius() : float{
+		return $this->initialRadius;
+	}
+
+	/**
+	 * Sets the initial radius.
+	 *
+	 * @throws \RuntimeException if something attempted to set when it is already ticking
```

Comment:
I don't like this. If the radius needs to be set in advance, it should be passed in the constructor.